geth/transactions/txqueue_manager_test.go

```diff
@@ -286,67 +278,113 @@ func (s *TxQueueTestSuite) TestInvalidPassword() {
 	})
 
-	s.NoError(txQueueManager.QueueTransaction(tx))
-
-	_, err := txQueueManager.CompleteTransaction(tx.ID, password)
+	s.NoError(s.manager.QueueTransaction(tx))
+	_, err := s.manager.CompleteTransaction(tx.ID, password)
 	s.Equal(err.Error(), keystore.ErrDecrypt.Error())
 
 	// Transaction should stay in the queue as mismatched accounts
 	// is a recoverable error.
-	s.True(txQueueManager.TransactionQueue().Has(tx.ID))
+	s.True(s.manager.TransactionQueue().Has(tx.ID))
 }
 
 func (s *TxQueueTestSuite) TestDiscardTransaction() {
-	txQueueManager := NewManager(s.nodeManagerMock, s.accountManagerMock)
-	txQueueManager.DisableNotificactions()
-
-	txQueueManager.Start()
-	defer txQueueManager.Stop()
-
 	tx := common.CreateTransaction(context.Background(), common.SendTxArgs{
 		From: common.FromAddress(TestConfig.Account1.Address),
 		To:   common.ToAddress(TestConfig.Account2.Address),
 	})
 
-	s.NoError(txQueueManager.QueueTransaction(tx))
+	s.NoError(s.manager.QueueTransaction(tx))
 	w := make(chan struct{})
 	go func() {
-		s.NoError(txQueueManager.DiscardTransaction(tx.ID))
+		s.NoError(s.manager.DiscardTransaction(tx.ID))
 		close(w)
 	}()
 
-	rst := txQueueManager.WaitForTransaction(tx)
+	rst := s.manager.WaitForTransaction(tx)
 	s.Equal(ErrQueuedTxDiscarded, rst.Error)
 	// Transaction should be already removed from the queue.
-	s.False(txQueueManager.TransactionQueue().Has(tx.ID))
+	s.False(s.manager.TransactionQueue().Has(tx.ID))
 	s.NoError(WaitClosed(w, time.Second))
 }
 
 func (s *TxQueueTestSuite) TestCompletionTimedOut() {
-	txQueueManager := NewManager(s.nodeManagerMock, s.accountManagerMock)
-	txQueueManager.DisableNotificactions()
-	txQueueManager.completionTimeout = time.Nanosecond
+	tx := common.CreateTransaction(context.Background(), common.SendTxArgs{
+		From: common.FromAddress(TestConfig.Account1.Address),
+		To:   common.ToAddress(TestConfig.Account2.Address),
+	})
 
-	txQueueManager.Start()
-	defer txQueueManager.Stop()
+	s.NoError(s.manager.QueueTransaction(tx))
+	rst := s.manager.WaitForTransaction(tx)
+	s.Equal(ErrQueuedTxTimedOut, rst.Error)
+}
 
+// TestLocalNonce verifies that local nonce will be used unless
+// upstream nonce is updated and higher than a local
+// in test we will run 3 transaction with nonce zero returned by upstream
+// node, after each call local nonce will be incremented
+// then, we return higher nonce, as if another node was used to send 2 transactions
+// upstream nonce will be equal to 5, we update our local counter to 5+1
+// as the last step, we verify that if tx failed nonce is not updated
+func (s *TxQueueTestSuite) TestLocalNonce() {
+	txCount := 3
+	password := TestConfig.Account1.Password
+	key, _ := crypto.GenerateKey()
+	account := &common.SelectedExtKey{
+		Address:    common.FromAddress(TestConfig.Account1.Address),
+		AccountKey: &keystore.Key{PrivateKey: key},
+	}
+	// 5 transactions in total
+	for i := 0; i < txCount+2; i++ {
+		s.setupStatusBackend(account, password, nil)
```

Comment:
It's a bit strange we call `s.setupStatusBackend(account, password, nil)` five times with the same args. Maybe we can clarify this in the comment: `setup expectations for 5 transactions in total`?